jobs/kube-apiserver/templates/config/cloud-provider.ini.erb

```diff
@@ -43,6 +43,20 @@
 
         workspace_config_string="[Workspace]\n"
         cloud_provider.if_p('cloud-provider.vsphere.resourcepool-path') { |path| workspace_config['resourcepool-path'] = path }
+      elsif provider_type == 'azure'
+        cloud_config['cloud'] = cloud_provider.p('cloud-provider.azure.cloud')
+        cloud_config['tenantId'] = cloud_provider.p('cloud-provider.azure.tenant-id')
+        cloud_config['aadClientId'] = cloud_provider.p('cloud-provider.azure.service-principal-id')
```

Comment:
@andyliuliming do we need to provide service-principal ID and secret for workers as well?
If yes, can we limit credentials for worker to some very specific minimal actions?

Reply:
yes, the principal id is created by customer. the customer can restrict the rights of it to only the resource groups or any resource it required only.